### PR TITLE
cake 5: Add native type hints for Http

### PR DESCRIPTION
**Pull request by othercorey**
@markstory Please take a close look at the changes to Cookie. We expect `array|string` values but have been accepting other values due to implicit string conversion when writing the headers.

### CODE REVIEW

File: src/Http/Cookie/Cookie.php [hunk 1]
```diff
@@ -416,15 +417,15 @@ protected function validateName(string $name): void
...
+    public function getValue(): array|string
     {
         return $this->value;
     }
 
     /**
      * @inheritDoc
      */
-    public function getScalarValue()
+    public function getScalarValue(): array|string
```
Comment (othercorey) [inline src/Http/Cookie/Cookie.php:428]: This one doesn't make much sense. The value is still `array|string` not `mixed` and certainly isn't converting anything.

File: tests/TestCase/Http/ResponseTest.php [hunk 2]
```diff
@@ -749,26 +749,6 @@ public function testWithCookieEmpty(): void
         $this->assertEquals($expected, $result);
     }
 
-    /**
-     * Test setting cookies with scalar values
-     */
-    public function testWithCookieScalar(): void
```
Comment (othercorey) [inline tests/TestCase/Http/ResponseTest.php:755]: These tests don't make much sense. They are testing the value stored in the Cookie object which is not the value sent and retrieved via headers. Those are expected to be strings.

File: src/Http/Cookie/Cookie.php [hunk 3]
```diff
@@ -146,7 +147,7 @@ class Cookie implements CookieInterface
      */
     public function __construct(
         string $name,
-        $value = '',
+        array|string $value = '',
```
[the remark this answers is not in the document]
Comment (othercorey) [reply src/Http/Cookie/Cookie.php:150]: I agree that numbers are probably used very often. Do we want just `int` or `float|int` ?

File: src/Http/Cookie/Cookie.php (hunk 1 above)
Comment (othercorey) [reply src/Http/Cookie/Cookie.php:428]: I was returning mixed. I'm not sure if static analyzer just ignore undefined return types though.

File: src/Http/Cookie/Cookie.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (ADmad) [reply src/Http/Cookie/Cookie.php:150]: `float|int`

File: src/Http/Cookie/Cookie.php (hunk 1 above)
Comment (ADmad) [reply src/Http/Cookie/Cookie.php:428]: The method calls `_flatten()` if the value is expanded (aka an array) which has return type `string`, so the return type here should should be a `string`.

File: src/Http/Cookie/Cookie.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (markstory) [reply src/Http/Cookie/Cookie.php:150]: We're unlikely to get floats from HTTP responses as that is all strings, but users creating cookies will use all sorts of scalar values. I agree that including `float|int` is a good idea, I might also add `bool` to that union, so we can cover all the useful primitives.

File: src/Http/Cookie/Cookie.php [hunk 4]
```diff
@@ -215,7 +216,7 @@ public static function setDefaults(array $options): void
      * @return static
      * @see \Cake\Cookie\Cookie::setDefaults()
      */
-    public static function create(string $name, $value, array $options = [])
+    public static function create(string $name, array|string $value, array $options = []): static
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php:219]: You might want `int|float|bool` here too.

File: src/Http/Cookie/Cookie.php [hunk 5]
```diff
@@ -451,7 +452,7 @@ public function withValue($value)
      * @param array|string $value The value to store.
      * @return void
      */
-    protected function _setValue($value): void
+    protected function _setValue(array|string $value): void
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php:455]: Would this need to take float|int|bool too?

File: src/Http/Cookie/CookieInterface.php [hunk 6]
```diff
@@ -79,24 +82,24 @@ public function getName(): string;
...
     /**
      * Gets the cookie value as scalar.
      *
      * This will collapse any complex data in the cookie with json_encode()
      *
-     * @return mixed
+     * @return array|string
      */
-    public function getScalarValue();
+    public function getScalarValue(): array|string;
```
Comment (markstory) [inline src/Http/Cookie/CookieInterface.php:94]: Should this be string?

File: src/Http/Cookie/CookieInterface.php [hunk 7]
```diff
@@ -79,24 +82,24 @@ public function getName(): string;
...
 
     /**
      * Create a cookie with an updated value.
      *
      * @param array|string $value Value of the cookie to set
      * @return static
      */
-    public function withValue($value);
+    public function withValue(array|string $value): static;
```
Comment (markstory) [inline src/Http/Cookie/CookieInterface.php:102]: Could add the `int|float|bool` here too. We could also narrow the type of value to only accept string|array, but I think accepting wider types and casting values to strings makes a more ergonomic API that is aligned with how CakePHP should work.

File: src/Http/Cookie/CookieInterface.php [hunk 8]
```diff
@@ -135,14 +138,14 @@ public function getDomain(): string;
...
-    public function withDomain(string $domain);
+    public function withDomain(string $domain): static;
 
     /**
      * Get the current expiry time
      *
      * @return \DateTime|\DateTimeImmutable|null Timestamp of expiry or null
      */
-    public function getExpiry();
+    public function getExpiry(): DateTime|DateTimeImmutable|null;
```
Comment (markstory) [inline src/Http/Cookie/CookieInterface.php:148]: Should this be DateTimeInterface?

File: src/Http/ServerRequest.php [hunk 9]
```diff
@@ -834,7 +834,7 @@ public function getHeader($name): array
      * @return string Header values collapsed into a comma separated string.
      * @link http://www.php-fig.org/psr/psr-7/ This method is part of the PSR-7 server request interface.
      */
-    public function getHeaderLine($name): string
```
Comment (markstory) [inline src/Http/ServerRequest.php:837]: Is this for compatibility with PSR interfaces?

File: src/Http/Cookie/Cookie.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (othercorey) [reply src/Http/Cookie/Cookie.php:150]: Converting `false` to `string` creates an empty string. Will that break the header or cookie?

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Http/Cookie/Cookie.php:150]: AFAIK there's nothing that prevents setting empty string as value for a cookie either, so `false` being converted to empty string shouldn't be an issue.

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Http/Cookie/Cookie.php:150]: `setcookie()` also has default value for `$value` arg as empty string.

[the remark this answers is not in the document]
Comment (markstory) [reply src/Http/Cookie/Cookie.php:150]: Empty strings in cookies are totally fine.

File: src/Http/ServerRequest.php (hunk 9 above)
Comment (othercorey) [reply src/Http/ServerRequest.php]: I don't know why this changed. I reset it to `string`.

**Pull request merged**